jgrapht-core/src/main/java/org/jgrapht/alg/BidirectionalDijkstraShortestPath.java

```diff
@@ -0,0 +1,490 @@
+        }
+        this.heap = new FibonacciHeap<>();
+        this.seen = new HashMap<>();
+    }
+
+    public void updateDistance(V v, E e, double distance)
+    {
+        FibonacciHeapNode<QueueEntry<V, E>> node = seen.get(v);
+        if (node == null) {
+            node = new FibonacciHeapNode<>(new QueueEntry<>(e, v));
+            heap.insert(node, distance);
+            seen.put(v, node);
+        } else {
+            if (distance < node.getKey()) {
+                heap.decreaseKey(node, distance);
+                node.getData().e = e;
+            }
+        }
+    }
+
+    public double getDistance(V v)
+    {
+        FibonacciHeapNode<QueueEntry<V, E>> node = seen.get(v);
+        if (node == null) {
+            return Double.POSITIVE_INFINITY;
+        } else {
+            return node.getKey();
+        }
+    }
+
+    public E getTreeEdge(V v)
+    {
+        FibonacciHeapNode<QueueEntry<V, E>> node = seen.get(v);
+        if (node == null) {
+            return null;
+        } else {
+            return node.getData().e;
+        }
+    }
+}
+
+/**
+ * Provides unified interface for operations that are different in directed
+ * graphs and in undirected graphs.
+ */
+abstract class Specifics<VV, EE>
+{
+
+    /**
+     * Returns the edges outgoing from the specified vertex in case of directed
+     * graph, and the edge touching the specified vertex in case of undirected
+     * graph.
+     *
+     * @param vertex the vertex whose outgoing edges are to be returned.
+     *
+     * @return the edges outgoing from the specified vertex in case of directed
+     *         graph, and the edge touching the specified vertex in case of
+     *         undirected graph.
+     */
+    public abstract Set<? extends EE> edgesOf(VV vertex);
+}
+
+/**
+ * An implementation of {@link Specifics} for a directed graph.
+ */
+class DirectedSpecifics<VV, EE>
+    extends Specifics<VV, EE>
+{
+
+    private DirectedGraph<VV, EE> graph;
+
+    /**
+     * Creates a new DirectedSpecifics object.
+     *
+     * @param g the graph for which this specifics object to be created.
+     */
+    public DirectedSpecifics(DirectedGraph<VV, EE> g)
+    {
+        graph = g;
+    }
+
+    /**
+     * @see CrossComponentIterator.Specifics#edgesOf(Object)
+     */
+    @Override
+    public Set<? extends EE> edgesOf(VV vertex)
+    {
+        return graph.outgoingEdgesOf(vertex);
+    }
+}
+
+/**
+ * An implementation of {@link Specifics} in which edge direction (if any) is
+ * ignored.
+ */
+class UndirectedSpecifics<VV, EE>
+    extends Specifics<VV, EE>
+{
+
+    private Graph<VV, EE> graph;
+
+    /**
+     * Creates a new UndirectedSpecifics object.
+     *
+     * @param g the graph for which this specifics object to be created.
+     */
+    public UndirectedSpecifics(Graph<VV, EE> g)
+    {
+        graph = g;
+    }
+
+    /**
+     * @see CrossComponentIterator.Specifics#edgesOf(Object)
```

Comment:
Not sure why you reference to the CrossComponentIterator here? I think that the qualifier "import org.jgrapht.traverse.CrossComponentIterator;" is also missing.